annotator-core/src/main/java/edu/ucr/cs/riple/core/evaluators/CachedEvaluator.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 Nima Karimipour
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package edu.ucr.cs.riple.core.evaluators;
+
+import static java.util.function.UnaryOperator.identity;
+import static java.util.stream.Collectors.toMap;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import edu.ucr.cs.riple.core.Report;
+import edu.ucr.cs.riple.core.cache.Impact;
+import edu.ucr.cs.riple.core.cache.TargetModuleCache;
+import edu.ucr.cs.riple.core.evaluators.suppliers.Supplier;
+import edu.ucr.cs.riple.core.metadata.index.Error;
+import edu.ucr.cs.riple.core.metadata.index.Fix;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** This evaluator uses modeling techniques to compute effect of applying a fix tree. */
```

Comment:
```suggestion
/** This evaluator uses caching techniques to optimize computing the effect of applying a fix tree. */
```

Reply:
In either version, this is a very self-referential javadoc for what is a very tricky class. You are basically saying "a CachedEvaluator evaluates the effect of fixes using a cache" (or modeling, which doesn't mean anything specific either). But it's the interaction between this class and `TargetModuleCache` that gives you the core optimization implemented in this PR, no? So they _both_ need better documentation.

As I understand it, the core difference between this and the existing `BasicEvaluator` is that we are only adding fixes to the conflict graph if `cache::isUnknown` returns `true` for that fix. It is then the presence of elements in that conflict graph which triggers any re-runs of the underlying NullAway compilation. Correct?

Then, question: when does `cache::isUnknown` happen? And I thought even without `CachedEvaluator` we were avoiding retrying the same fixes if we already knew their impact. Is that wrong? What kind of results are we caching now that we weren't before?

 I think it would be worth adding a more detailed javadoc not only here, but also on `CachedEvaluator::initializeFixGraph` and `collectGraphResults`. Usually, an `@Override` method doesn't need its own Javadoc if it's obvious how it's implementing its superclass' method, but here we are very significantly changing what gets computed by these methods. No?

In general, there _are_ cases where we the javadoc comments are going to be short and provide little information over the method name (e.g. `getIterator()`: "get the iterator for this collection" is fine), but if you find yourself writing quick formulaic javadocs for complicated code that only mean something if you already have the full algorithm/architecture in your head, then that's usually a sign that better documentation is needed.

Reply:
Thank you for all the notes and all the suggestions, I realized how incomplete the existing javadocs are! (mistakenly assumed the algorithm should only be discussed in the paper) I updated javadoc 0e568cb593d854a69822e638aaeb0623918a6357 and ad0da81b8de51c8bd0f33f3667576047308aef70.

The main difference between `BasicEvaluator` and `CachedEvaluator` is that `BasicEvaluator` includes the entire tree (with no exclusion) in the conflict graphs node and it does not store or use any impacts from previous iterations. `CachedEvaluator` only processes nodes that are not stored in the cache and assigns a single fix to each node.

Reply:
@lazaroclapp Please let me know if I should add anything else.